\documentclass[thesis.tex]{subfiles}

\begin{document}
\chapter{Method}
In this chapter we will present an algorithm for finding guaranteed optimal alignments for arbitrary reads against a graph-based reference genome given an error threshold. The notion of optimal will be defined in a more detailed specification of the problem alongside a rigorous description of the graphs involved. The PO-MSA algorithm briefly mentioned in the background has been used as a baseline in the experiments so an accurate description of the implementation of both this and the scoring matrix used is included.
\subsection{Scoring schema}
The algorithm is general enough to work with any scoring schema that yields a scoring function which is monotone in relation to the length of the input sequence. For these experiments a variant of the Levensthein distance schema is used (as seen in \ref{fig:edit_distance_variant}), with the extension of penalizing both opening and extending a gap with -1. The impact of nonincreasing vs nondecreasing scoring functions is discussed \textcolor{red}{SOMEWHERE IN THE DISCUSSION}.
\begin{wrapfigure}{r}{0.4\textwidth}
	$
	\begin{array}{r|rrrr}
	 & \mathbf{A} & \mathbf{C} & \mathbf{G} & \mathbf{T} \\ \hline
	\mathbf{A} & 0 & -1 & -1 & -1 \\
	\mathbf{C} & -1 & 0 & -1 & -1 \\
	\mathbf{G} & -1 & -1 & 0 & -1 \\
	\mathbf{T} & -1 & -1 & -1 & 0
	\end{array}
	$
	\label{fig:edit_distance_variant}
	\caption{The scoring matrix used in the experiments}
\end{wrapfigure}
\subsection{Data}
The reference data used in testing is real test data borrowed from Erik Garrisons project vg\ref{REFERENCE} and the tool developed from the article \textit{Canonical, Stable, General Mapping using Context Schemes}\ref{canonical_stable_general_mapping}. Read data is generated by taking random sequences from the reference graph and introducing noise. When not specified the length of the original sequences are 60 and the probabilities of introducing a SNP, an insert or a deletion is 0.01 for each.
\subsection{PO-MSA implementation}
\subsection{Validation}
In the results reads are categorized as either correctly or incorrectly aligned against the reference. This separation is made on based on whether the produced alignment has an alignment score higher than or equal to the alignment produced by PO-MSA. As PO-MSA is guaranteed to find an optimal score, results where the algorithm finds a better alignment then PO-MSA is interpreted as an error either in the implementation of PO-MSA or the scoring process. 
\end{document}
